oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/ExternalGroupPrincipalProvider.java

```diff
@@ -0,0 +1,353 @@
+ * {@link org.apache.jackrabbit.oak.spi.security.authentication.external.ExternalIdentityProvider}.
+ *
+ * For performance reasons this implementation doesn't lookup principals on the IDP
+ * but relies on a persisted cache inside the repository where the names of these
+ * external principals are synchronized to based on a configurable expiration time.
+ *
+ * Currently, the implementation respects the {@code rep:externalPrincipalNames}
+ * properties, where group membership of external users gets synchronized if
+ * {@link DefaultSyncConfig.User#getDynamicMembership() dynamic membership} has
+ * been enabled. This synchronization is take care of by the {@link PrincipalSyncContext}.
+ *
+ * Please note that in contrast to the default principal provider implementation
+ * shipped with Oak the group principals known and exposed by this provider are
+ * not backed by a authorizable group and thus cannot be retrieved using
+ * Jackrabbit user management API.
+ */
+class ExternalGroupPrincipalProvider implements PrincipalProvider, ExternalIdentityConstants {
+
+    private static final Logger log = LoggerFactory.getLogger(ExternalGroupPrincipalProvider.class);
+
+    private final Root root;
+    private final NamePathMapper namePathMapper;
+    private final UserManager userManager;
+
+    ExternalGroupPrincipalProvider(Root root, UserConfiguration uc, NamePathMapper namePathMapper) {
+        this.root = root;
+        this.namePathMapper = namePathMapper;
+        userManager = uc.getUserManager(root, namePathMapper);
+    }
+
+    //--------------------------------------------------< PrincipalProvider >---
+    @Override
+    public Principal getPrincipal(@Nonnull String principalName) {
+        Result result = findPrincipals(principalName, true);
+        if (result != null && result.getRows().iterator().hasNext()) {
+            return new ExternalGroupPrincipal(principalName);
+        } else {
+            return null;
+        }
+    }
+
+    @Nonnull
+    @Override
+    public Set<Group> getGroupMembership(@Nonnull Principal principal) {
+        if (!(principal instanceof Group)) {
+            try {
+                if (principal instanceof ItemBasedPrincipal) {
+                    Tree t = root.getTree(((ItemBasedPrincipal) principal).getPath());
+                    return getGroupPrincipals(t);
+                } else {
+                    return getGroupPrincipals(userManager.getAuthorizable(principal));
+                }
+            } catch (RepositoryException e) {
+                log.debug(e.getMessage());
+            }
+
+        }
+        return ImmutableSet.of();
+    }
+
+    @Nonnull
+    @Override
+    public Set<? extends Principal> getPrincipals(@Nonnull String userID) {
+        try {
+            return getGroupPrincipals(userManager.getAuthorizable(userID));
+        } catch (RepositoryException e) {
+            log.debug(e.getMessage());
```

Comment:
not recommended to use the message as is here. better: `log.debug("{}", e.getMessage());`